core/node/api_server/src/execution_sandbox/apply.rs

```diff
@@ -37,179 +40,250 @@ use zksync_types::{
-            L2BlockEnv {
-                number: 1,
-                timestamp: 0,
-                prev_block_hash: L2BlockHasher::legacy_hash(L2BlockNumber(0)),
-                max_virtual_blocks_to_create: 1,
-            }
-        } else {
-            // We need to reset L2 block info in storage to process transaction in the current block context.
-            // Actual resetting will be done after `storage_view` is created.
-            let prev_l2_block_info = StoredL2BlockInfo::new(
-                connection,
-                resolved_block_info.state_l2_block_number - 1,
-                None,
-            )
+    let next_block = if is_pending_block {
+        L2BlockEnv {
+            number: next_block.number + 1,
+            timestamp: resolved_block_info.l1_batch_timestamp,
+            prev_block_hash: resolved_block_info.state_l2_block_hash,
+            // For simplicity, we assume each L2 block create one virtual block.
+            // This may be wrong only during transition period.
+            max_virtual_blocks_to_create: 1,
+        }
+    } else if next_block.number == 0 {
+        // Special case:
+        // - For environments, where genesis block was created before virtual block upgrade it doesn't matter what we put here.
+        // - Otherwise, we need to put actual values here. We cannot create next L2 block with block_number=0 and `max_virtual_blocks_to_create=0`
+        //   because of SystemContext requirements. But, due to intrinsics of SystemContext, block.number still will be resolved to 0.
+        L2BlockEnv {
+            number: 1,
+            timestamp: 0,
+            prev_block_hash: L2BlockHasher::legacy_hash(L2BlockNumber(0)),
+            max_virtual_blocks_to_create: 1,
+        }
+    } else {
+        // We need to reset L2 block info in storage to process transaction in the current block context.
+        // Actual resetting will be done after `storage_view` is created.
+        let prev_block_number = resolved_block_info.state_l2_block_number - 1;
+        let prev_l2_block = read_stored_l2_block(connection, prev_block_number)
             .await
             .context("failed reading previous L2 block info")?;
 
-            l2_block_info_to_reset = Some(prev_l2_block_info);
-            L2BlockEnv {
-                number: current_l2_block_info.l2_block_number,
-                timestamp: current_l2_block_info.l2_block_timestamp,
-                prev_block_hash: prev_l2_block_info.l2_block_hash,
-                max_virtual_blocks_to_create: 1,
-            }
+        let mut prev_block_hash = connection
+            .blocks_web3_dal()
+            .get_l2_block_hash(prev_block_number)
+            .await
+            .map_err(DalError::generalize)?;
+        if prev_block_hash.is_none() {
+            // We might need to load the previous block hash from the snapshot recovery metadata
```

Comment:
This was incorrectly not done previously, which led to the RPC method erroring if asked the first retained L2 block. Again, this wasn't covered by tests because the mock oneshot executor returned before this code has a chance to run.